src/Simulation/Simulators/OpenSystemsSimulator/NativeInterface.cs

```diff
@@ -1,265 +1,302 @@
 
 #nullable enable
 
-namespace Microsoft.Quantum.Experimental
+namespace Microsoft.Quantum.Experimental;
+
+/// <summary>
+///     Represents an exception that is raised by native simulator code.
+/// </summary>
+[Serializable()]
+public class SimulationException : Exception
 {
+    private readonly string? source;
+    private readonly string? backtrace;
+    private readonly string message;
+
     /// <summary>
-    ///     Represents an exception that is raised by native simulator code.
+    ///     The name of the native shared library which raised this
+    ///     exception if known, <c>null</c> otherwise.
     /// </summary>
-    [Serializable()]
-    public class SimulationException : Exception
+    public string? SourceLibrary => source;
+
+    /// <summary>
+    ///      The native backtrace, if any, associated with this exception.
+    /// </summary>
+    public string? NativeBacktrace => backtrace;
+
+    public override string Message =>
+        $"Exception in native open systems simulator runtime {(source is null ? "" : $"({source})")}: {this.message}" +
+        (backtrace is null
+         ? ""
+         : $"\nNative backtrace:\n{NativeBacktrace}");
+
+    internal SimulationException(string message, string? backtrace, string? source) : base(message)
     {
-        private readonly string? source;
+        this.message = message;
+        this.backtrace = backtrace;
+        this.source = source;
+    }
+}
 
-        /// <summary>
-        ///     The name of the native shared library which raised this
-        ///     exception if known, <c>null</c> otherwise.
-        /// </summary>
-        public string? SourceLibrary => source;
+/// <summary>
+///     Abstracts away calls to and from the experimental simulators DLL.
+/// </summary>
+internal static class NativeInterface
+{
+    public static event Action<string>? OnVerbose = null;
+    private static void LogCall(string callName) =>
+        OnVerbose?.Invoke($"[VERBOSE] NativeInterface: calling {callName}.");
 
-        internal SimulationException(string message, string? source) : base(message)
+    private static void CheckCall(Int64 errorCode)
+    {
+        if (errorCode != 0)
         {
-            this.source = source;
+            var error = _LastError();
+            var bt = _LastBackTrace();
+            throw new SimulationException(error, bt, DLL_NAME);
         }
     }
 
-    /// <summary>
-    ///     Abstracts away calls to and from the experimental simulators DLL.
-    /// </summary>
-    internal static class NativeInterface
+    public const string DLL_NAME = "Microsoft.Quantum.Experimental.Simulators.Runtime.dll";
+
+    [DllImport(DLL_NAME, ExactSpelling=true, CallingConvention=CallingConvention.Cdecl, EntryPoint="lasterr")]
+    private static extern string _LastError();
+
+    [DllImport(DLL_NAME, ExactSpelling=true, CallingConvention=CallingConvention.Cdecl, EntryPoint="lastbacktrace")]
+    private static extern string? _LastBackTrace();
```

Comment:
(Subjective judgement)
Passing nullable value (or a reference?) across DLL boundary makes me somewhat uncertain... I assume that _nullability_ can be implemented differently on different sides of the DLL boundary. And class instances (`string`), as opposed to primitive types, can also be implemented differently...

Reply:
Here, the `DllImport` attribute causes the .NET runtime to use P/Invoke in marshalling data across the managed/unmanaged DLL boundary — in particular, I believe that P/Invoke should marshal `char* result = nullptr` as a managed null reference of type `System.String`. Since in the .NET runtime, all reference types are nullable, nullability data (that is, `string` vs `string?`) should only affect nullability metadata used by the compiler for type checking at the C# level, such that I don't think there's any issues with it being implemented differently.

That is, if this line was changed to `private static extern string _LastBackTrace()`, the emitted CIL would differ only by nullability metadata used at compile-time when linking against this DLL.